prow/plank/reconciler.go

```diff
@@ -333,8 +333,10 @@ func (r *reconciler) syncPendingJob(ctx context.Context, pj *prowv1.ProwJob) (*r
 	prevPJ := pj.DeepCopy()
 
 	pod, podExists, err := r.pod(ctx, pj)
-	if err != nil {
-		return nil, err
+	if time.Since(pod.CreationTimestamp.Time) >= r.config().Plank.PodUnscheduledTimeout.Duration {
```

Comment:
We shouldn't be conditionally error checking and I don't understand how this helps with the linked issue, could you add a unittest that demonstrates the issue and fix, please?

Reply:
@alvaroaleman as mentioned in the issue description that the code should not return error right away when the build cluster is unreachable, it should also check pod_unscheduled_timeout, so I added this condition.

Please inform if this change is wrong, if yes what should be the right approach?

Reply:
Please add a test that reproduces the bug described in the linked issue together with the fix.

Also, error checking has to always happen, so nesting it inside another check is not an option, please find another approach.

Reply:
An obvious problem I'm seeing here, it that the first returned value of `r.pod(ctx, pj)` is a pointer, it could be nil and the current implementation would seg fault.

Some thoughts around solving this problem:
- If the error happened here, there could be two cases:
  - a). the pod doesn't exist. In this case we can say we want to respect the unschedule timeout
  - b). the pod exist but build cluster api temporarily unavailable. In this case it's correct to return the error and let the caller decide what to do

So I agree with @alvaroaleman  that the error should not be checked optionally to avoid breaking b).

one idea here is to use `IsTerminalError` function here to decide what to do, for ref https://github.com/NikhilSharmaWe/test-infra/blob/5578bc72e6677431e0326675c699d7ef6b0fc2ec/prow/plank/reconciler.go#L272